src/index_queue.rs

```diff
@@ -0,0 +1,273 @@
+        // which will give us a list of CIDs that contain the keyword.
+
+        if self.keywords.contains_key(&query) {
+            let cid_map: CHashMap<String, ()> = self.keywords.get(&query).unwrap().clone();
+            cid_map.into_iter().for_each(|(cid, _)| {
+                if self.map.contains_key(&cid) {
+                    let index_result: IndexResult = self.map.get(&cid).unwrap().clone();
+                    results.push(index_result);
+                }
+            });
+        }
+
+        return results;
+    }
+
+    pub fn start(&self, gateway: String) {
+        loop {
+            if self.queue.is_empty() == false {
+                let try_item = self.queue.pop();
+                if try_item.is_some() {
+                    let item = try_item.unwrap();
+                    self.queue_set.remove(&*item);
+                    info!("Indexing {}", item);
+
+                    let result = self.retreive_content(gateway.clone(), item.clone());
+
+                    if result.is_some() {
+                        self.map.insert(item.clone(), result.unwrap());
+                    } else {
+                        info!("Error retrieving CID {}", item);
+                        // self.enqueue(item.clone()); // for now give up on error
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Use the http client to obtain the page from the ipfs gateway. If there is a failure to
+     * obtain the CID, we give up for now.
+     */
+    fn retreive_content(&self, gateway: String, cid: String) -> Option<IndexResult> {
+        let url = format!("http://{}/ipfs/{}", gateway, cid);
+        info!("Retreiving {}", url);
+        let client = reqwest::blocking::Client::new();
+        let result = client.get(&url).send();
+        let result = match result {
+            Ok(result) => result,
+            Err(err) => {
+                info!("Error: {}, not re-enqueue-ing cid", err);
+                // self.enqueue(cid.clone()); // for now give up on error
+                return None;
+            }
+        };
+        let response = result;
+
+        // todo: check the file type header and only proceed this way if it is actually html, otherwise index as some other file type
+        // plus some meta data
+        let html = response.text().unwrap();
+        let mut document = Html::parse_document(html.as_str());
+        let result = self.detect_redirect(url, cid.clone(), document.clone());
+        let mut fullcid = cid.clone();
+        if result.is_some() {
+            (fullcid, document) = result.unwrap();
+        }
+        trace!("received: {:?}", html.as_str());
+
+        self.process_content(gateway, fullcid.clone(), document.clone())
+    }
+
+    /**
+     * Determine if the response we've received requires another request from a redirect.
+     * If it requires another request, we will have an updated "full cid" as well.
```

Comment:
Is "full cid" a naming scheme that's specified somewhere that I could reference?

Reply:
nah its just an internal thing - prob could use a better name, its just the cid + the path. so for example wikipedia might have the hash `QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco` but you get redirected to the page: `QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/wiki` so what I've been calling the fullcid is is the hash + `/wiki`